win32/src/machine_emu.rs

```diff
@@ -145,91 +142,134 @@ impl MachineX<Emulator> {
         }
     }
 
-    pub fn run(&mut self) -> bool {
+    pub fn run(&mut self, instruction_count: usize) -> StopReason {
         self.emu.x86.schedule();
+        let eip = self.emu.x86.cpu().regs.eip;
         match &self.emu.x86.cpu().state {
-            x86::CPUState::Running => self.execute_block(),
+            x86::CPUState::Running => self.execute_block(instruction_count),
             x86::CPUState::Blocked(wait) => {
                 let wait = *wait;
                 if self.host.block(wait) {
                     self.unblock();
-                } else {
-                    return false;
                 }
+                StopReason::Blocked
+            }
+            x86::CPUState::DebugBreak => {
+                // resume execution after breakpoint
+                self.emu.x86.cpu_mut().state = x86::CPUState::Running;
+                self.execute_block(instruction_count)
             }
-            _ => return false,
+            x86::CPUState::Error(message) => StopReason::Error {
```

Comment:
BTW one reason I had a weird API here around returning a bool is that this run function is the hottest code when emulating.  So when this returns a constructed StopReason enum, the caller is responsible for tearing that StopReason down when it drops which means the drop impl needs to check if it's a StopReason::Error and if so free the message.  I cannot remember if I measured this mattering or if I was just worrying about it in the abstract.

Reply:
I'm not sure the drop check here will matter much, it won't have to do anything in most cases. Happy to rework it if you have suggestions, though